### PR TITLE
RFI - Support building System.Drawing for NETCOREAPP2.0

### PR DESCRIPTION
**Pull request by qmfrederik**
This PR enables building System.Drawing as a netcoreapp2.0 library; together with libgdiplus it allows those who want to, to use System.Drawing on .NET Core on Linux, macOS or Windows.

All it takes are a couple of `#if NETCOREAPP2_0` statements in the code base. I'm not sure if you have similar code already - happy to change that to whatever preprocessor variable you use to detect you're compiling for .NET Core.

There are some other changes in this PR that allow this to build without any warnings, I can undo those if you're not happy with them.

The [.csproj file](https://github.com/CoreCompat/System.Drawing/blob/master/src/CoreCompat.System.Drawing/CoreCompat.System.Drawing.csproj) which enables the final building is available elsewhere. If you think it makes sense to add it to Mono, too, I'll add it to this PR.

Let me know what you think.

### CODE REVIEW

File: mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/TestMatrix.cs [hunk 1]
```diff
@@ -40,8 +40,8 @@ namespace MonoTests.System.Drawing.Drawing2D
 	public class MatrixTest {
 
 		private Matrix default_matrix;
-		private Rectangle rect;
-		private RectangleF rectf;
+		private Rectangle rect = default (Rectangle);
```
Comment (marek-safar) [inline mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/TestMatrix.cs]: Why do you need this?

File: mcs/class/System.Drawing/System.Drawing/KnownColors.cs [hunk 2]
```diff
@@ -269,6 +269,7 @@ static void RetrieveWindowsSystemColors ()
 
 		public static Color FromKnownColor (KnownColor kc)
 		{
+#if !NETCOREAPP2_0
```
Comment (akoeplinger) [inline mcs/class/System.Drawing/System.Drawing/KnownColors.cs]: avoid the negative please, switching the conditions makes it clearer.

File: mcs/class/System.Drawing/System.Drawing/KnownColors.cs [hunk 3]
```diff
@@ -284,6 +285,9 @@ public static Color FromKnownColor (KnownColor kc)
 			}
 			c.knownColor = n;
 			return c;
+#else
+			return Color.FromKnownColor (kc);
```
Comment (akoeplinger) [inline mcs/class/System.Drawing/System.Drawing/KnownColors.cs]: We could instead move our implementation to `Color` as well so we don't need this ifdef at all.

File: mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/TestMatrix.cs (hunk 1 above)
Comment (qmfrederik) [inline mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/TestMatrix.cs]: Only purpose was to silence a compiler warning [CS0649 Field 'field' is never assigned to, and will always have its default value 'value'](https://msdn.microsoft.com/en-us/library/03b5270t.aspx). So not a strict requirement per se, I can revert this if you prefer.

File: mcs/class/System.Drawing/System.Drawing/KnownColors.cs (hunk 3 above)
Comment (qmfrederik) [inline mcs/class/System.Drawing/System.Drawing/KnownColors.cs]: Makes sense, I updated the PR.

File: mcs/class/System.Drawing/System.Drawing/Color.cs [hunk 4]
```diff
@@ -151,7 +151,21 @@ public static Color FromArgb (int argb)
 
 		public static Color FromKnownColor (KnownColor color)
 		{
-			return KnownColors.FromKnownColor (color);
+			Color c;
+			short n = (short)kc;
+			if ((n <= 0) || (n >= ArgbValues.Length)) {
```
Comment (akoeplinger) [inline mcs/class/System.Drawing/System.Drawing/Color.cs]: This won't compile (ArgbValues is in KnownColor).

File: mcs/class/System.Drawing/System.Drawing/Color.cs [hunk 5]
```diff
@@ -151,7 +151,21 @@ public static Color FromArgb (int argb)
...
+				// This is what it returns!
+				c = Color.FromArgb (0, 0, 0, 0);
+				c.state |= (short) Color.ColorType.Named;
+			} else {
+				c = new Color ();
+				c.state = (short) (Color.ColorType.ARGB | Color.ColorType.Known | Color.ColorType.Named);
+				if ((n < 27) || (n > 169))
+					c.state |= (short) Color.ColorType.System;
```
Comment (akoeplinger) [inline mcs/class/System.Drawing/System.Drawing/Color.cs]: all the Color. prefixes can now be removed.

File: mcs/class/System.Drawing/System.Drawing/Color.cs [hunk 6]
```diff
@@ -151,7 +151,21 @@ public static Color FromArgb (int argb)
...
+				c.Value = ArgbValues [n];
+			}
+			c.knownColor = n;
+			return c;
 		}
 
 		public static Color FromName (string name)
```
Comment (akoeplinger) [inline mcs/class/System.Drawing/System.Drawing/Color.cs]: This one uses KnownColors.FromKnownColor which could use FromKnownColor now directly.

File: mcs/class/System.Drawing/System.Drawing/Color.cs (hunk 5 above)
Comment (qmfrederik) [inline mcs/class/System.Drawing/System.Drawing/Color.cs]: OK, done.

File: mcs/class/System.Drawing/System.Drawing/Color.cs (hunk 6 above)
Comment (qmfrederik) [inline mcs/class/System.Drawing/System.Drawing/Color.cs]: OK, done.

File: mcs/class/System.Drawing/System.Drawing/Color.cs (hunk 4 above)
Comment (qmfrederik) [inline mcs/class/System.Drawing/System.Drawing/Color.cs]: Oops, fixed!

File: mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/TestMatrix.cs (hunk 1 above)
Comment (marek-safar) [inline mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/TestMatrix.cs]: Please remove these fields and inline their usage, it'll make the tests cleaner

Comment (qmfrederik) [inline mcs/class/System.Drawing/Test/System.Drawing.Drawing2D/TestMatrix.cs]: OK, done. Looks like default_matrix is an unused field as well, so I removed that, too.

**Comment by qmfrederik** [comment]
@akoeplinger @marek-safar Thx for the feedback. I had a quick look at the build, the failures seem unrelated.

Let me know if you need anything else from my side. Thx!

**Pull request merged**